molecules/predict.py

```diff
@@ -0,0 +1,180 @@
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+from __future__ import print_function
+
+import argparse
+import json
+import os
+import sys
+import tempfile
+
+import pubchem
+
+import apache_beam as beam
+import tensorflow as tf
+
+from apache_beam.options.pipeline_options import GoogleCloudOptions
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.options.pipeline_options import SetupOptions
+from apache_beam.options.pipeline_options import StandardOptions
+from tensorflow.python.framework import ops
+from tensorflow.python.saved_model import loader
+
+
+class Predict(beam.DoFn):
+  def __init__(self,
+      model_dir,
+      id_key,
+      meta_tag='serve',
+      meta_signature='predict',
+      meta_predictions='predictions'):
+    super(Predict, self).__init__()
+    self.model_dir = model_dir
+    self.id_key = id_key
+    self.meta_tag = meta_tag
+    self.meta_signature = meta_signature
+    self.meta_predictions = meta_predictions
+    self.session = None
+    self.graph = None
+    self.feed_tensors = None
+    self.fetch_tensors = None
+
+  def process(self, inputs):
+    # Create a session for every worker only once. The session is not
+    # pickleable, so it can't be created at the DoFn constructor.
+    if not self.session:
+      self.graph = ops.Graph()
+      with self.graph.as_default():
+        self.session = tf.Session()
+        metagraph_def = loader.load(
+            self.session, {self.meta_tag}, self.model_dir)
+      signature_def = metagraph_def.signature_def[self.meta_signature]
+
+      # inputs
+      self.feed_tensors = {
+          k: self.graph.get_tensor_by_name(v.name)
+          for k, v in signature_def.inputs.items()
+      }
+
+      # outputs/predictions
+      self.fetch_tensors = {
+          k: self.graph.get_tensor_by_name(v.name)
+          for k, v in signature_def.outputs.items()
+      }
+
+    # Create a feed_dict for a single element.
+    feed_dict = {
+        tensor: [inputs[key]]
+        for key, tensor in self.feed_tensors.items()
+        if key in inputs
+    }
+    results = self.session.run(self.fetch_tensors, feed_dict)
+
+    yield {
+        'id': inputs[self.id_key],
+        'predictions': results[self.meta_predictions][0].tolist()
+    }
+
+
+def run(model_dir, feature_extraction, sink, beam_options=None):
+  with beam.Pipeline(options=beam_options) as p:
+    _ = (p
+        | 'Feature extraction' >> feature_extraction
+        | 'Predict' >> beam.ParDo(Predict(model_dir, 'ID'))
+        | 'Format as JSON' >> beam.Map(lambda result: json.dumps(result))
+        | 'Write to sink' >> sink)
```

Comment:
Same technicality. In this line, `sink` is rather a `beam.Write(...)` PTransform that writes to some sink. It's no big deal, but maybe `sink_transform`, or `write_transform` or something like that would be more appropriate, since we want this to be a learning material.

Reply:
Done